### PR TITLE
Save chart configuration

### PR DESCRIPTION
**Pull request by IvanIvanoff**
### Create chart configuration
Create requires an input object with the settings:
```graphql
mutation {
  createChartConfiguration(
    settings: {
      title: "some title"
      description: "some description"
      isPublic: true
      anomalies: ["list", "of", "anomalies"]
      metrics: ["list", "of", "metrics"]
      projectId: 12}) {
        id
        title
        isPublic
        description
        user {
          id
          email
        }
        project {
          id
          slug
        }
        metrics
        anomalies
  }
}
```

### Update chart configuration
Update requires a chart configuration id and settings:
```graphql
mutation {
  updateChartConfiguration(
    id:1
    settings: {
      title: "some title"
      description: "some description"
      isPublic: true
      anomalies: ["list", "of", "anomalies"]
      metrics: ["list", "of", "metrics"]
      projectId: 12}) {
        id
       # ...
  }
}
```

### Delete a chart configuration
Delete requires only id of a chart configuration
```graphql
mutation {
  deleteChartConfiguration(id: 1) {
    id
  }
}
```

#### Single configuration
```graphql
{
  chartConfiguration(id: 1) {
    id
    # ...
  }
}
```

#### All accessible configurations (public + all private owned by the current user):
```graphql
{
  chartConfigurations {
    id
    # ...
  }
}
```

#### All accessible configurations for a given user (public + all private if the user_id is the same as the current user's id):
```graphql
{
  chartConfigurations(user_id: 1) {
    id
    # ...
  }
}
```

#### All accessible configurations for a project (public + all private owned by the current user):
```graphql
{
  chartConfigurations(project_id: 1) {
    id
    # ...
  }
}
```

#### All accessible configurations for of a user for a project (public + all private owned if the user_id is the id of the current user):
```graphql
{
  chartConfigurations(user_id: 12, project_id: 55) {
    id
    # ...
  }
}
```

[//]: # (#### Related PRs)

[//]: # (#### Additional deploy notes)

[//]: # (#### Screenshots)

### CODE REVIEW

File: lib/sanbase/auth/user_can.ex [hunk 1]
```diff
@@ -0,0 +1,29 @@
+defimpl Canada.Can, for: Sanbase.Auth.User do
```
Comment (tspenov) [inline lib/sanbase/auth/user_can.ex]: Is this used at all ?

Comment (IvanIvanoff) [reply lib/sanbase/auth/user_can.ex]: Not now, I'll remove it at the end when everything is ready and if it is still not used

File: lib/sanbase/chart/configuration.ex [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+        |> Repo.delete()
+
+      {:error, error} ->
+        {:error, "Cannot delete chart configuration. Reason: #{inspect(error)}"}
+    end
+  end
+
+  defp get_chart_configuration(conf_id, querying_user_id) do
+    case Repo.get(__MODULE__, conf_id) do
+      {:ok, %__MODULE__{user_id: user_id, is_public: is_public} = conf}
```
Comment (tspenov) [inline lib/sanbase/chart/configuration.ex]: `Repo.get` returns `%MODULE{} | nil` not :ok/:error tuple.
I think cleaner way will be with several function heads like:
```elixir
defp get_chart_configuration(conf_id, nil) do
  from(c in __MODULE__, where: c.id == ^conf_id and c.is_public == true) |> Repo.one
end

defp get_chart_configuration(conf_id, querying_user_id) when is_integer(querying_user_id) do
  from(c in __MODULE__, where: c.id == ^conf_id and c.user_id == ^querying_user_id) |> Repo.one
end
```

Comment (tspenov) [reply lib/sanbase/chart/configuration.ex]: And then maybe wrap it in :ok/error tuple

File: lib/sanbase/chart/configuration.ex [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+    project_chart_configurations_query(project_id, querying_user_id)
+    |> Repo.all()
+  end
+
+  def configurations(querying_user_id \\ nil) do
+    configurations_query(querying_user_id)
+    |> Repo.all()
+  end
+
+  def changeset(%__MODULE__{} = conf, attrs \\ %{}) do
```
Comment (tspenov) [inline lib/sanbase/chart/configuration.ex]: nitpick: can you move after the schema - I think we have in most schema modules

File: lib/sanbase/chart/configuration.ex [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+  schema "chart_configurations" do
+    field(:title, :string)
+    field(:description, :string)
+    field(:is_public, :boolean, default: false)
+
+    field(:metrics, {:array, :string}, default: [])
+    field(:anomalies, {:array, :string}, default: [])
+
+    belongs_to(:user, Sanbase.Auth.User)
+    belongs_to(:project, Sanbase.Model.Project)
```
Comment (tspenov) [inline lib/sanbase/chart/configuration.ex]: I think it is good to put the corresponding relations in the `User` and `Project`

File: lib/sanbase_web/graphql/schema/queries/project_chart_queries.ex [hunk 5]
```diff
@@ -0,0 +1,60 @@
+defmodule SanbaseWeb.Graphql.Schema.ProjectChartQueries do
+  @moduledoc ~s"""
+  Queries and mutations for working with Insights
```
Comment (tspenov) [inline lib/sanbase_web/graphql/schema/queries/project_chart_queries.ex]: some leftover

File: lib/sanbase_web/graphql/schema/queries/project_chart_queries.ex [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+  object :project_chart_mutations do
+    field :create_chart_configuration, :chart_configuration do
+      arg(:settings, :project_chart_input_object)
+
+      middleware(JWTAuth)
+      resolve(&ProjectChartResolver.create_chart_configuration/3)
+    end
+
+    field :update_chart_configuration, :chart_configuration do
+      arg(:id, non_null(:id))
```
Comment (tspenov) [inline lib/sanbase_web/graphql/schema/queries/project_chart_queries.ex]: :settings input object here :?

**Comment by tspenov** [comment]
Please add some tests

File: lib/sanbase/chart/configuration.ex (hunk 3 above)
Comment (IvanIvanoff) [reply lib/sanbase/chart/configuration.ex]: the schema is defined on line 8, this is line 41?

Comment (tspenov) [reply lib/sanbase/chart/configuration.ex]: Yes. I mean we put `changeset` func after `schema` definition (and before other public funcs) on most of such modules, so we can do that here as well.

Comment (IvanIvanoff) [reply lib/sanbase/chart/configuration.ex]: `can you move after the schema` I misunderstood this as if you're saying that now it's before the schema, will move it up

File: mix.exs [hunk 7]
```diff
@@ -110,7 +110,8 @@ defmodule Sanbase.Mixfile do
       {:inflex, "~> 2.0", override: true},
       {:mogrify, "~> 0.7.2"},
       {:lz4b, "0.0.4"},
-      {:rexbug, ">= 1.0.0", only: [:dev]}
+      {:rexbug, ">= 1.0.0", only: [:dev]},
+      {:canada, "~> 2.0"}
```
Comment (tspenov) [inline mix.exs]: leftover

File: lib/sanbase_web/graphql/schema/queries/chart_configuration_queries.ex [hunk 8]
```diff
@@ -0,0 +1,54 @@
+defmodule SanbaseWeb.Graphql.Schema.ChartConfigurationQueries do
+  @moduledoc ~s"""
+  Queries and mutations for working with Insights
+  """
+  use Absinthe.Schema.Notation
+
+  # import SanbaseWeb.Graphql.Cache, only: [cache_resolve: 1]
```
Comment (tspenov) [inline lib/sanbase_web/graphql/schema/queries/chart_configuration_queries.ex]: commented code

**Pull request merged**